components/dashboard/src/components/UsageBasedBillingConfig.tsx

```diff
@@ -116,75 +136,24 @@ export default function UsageBasedBillingConfig({ attributionId }: Props) {
-                window.localStorage.removeItem(localStorageKey);
-                clearTimeout(pollStripeSubscriptionTimeout!);
-                setPendingStripeSubscription(undefined);
+                setPendingStripeSubscription(attributionId, undefined);
                 setErrorMessage(`Could not subscribe to Stripe. ${error?.message || String(error)}`);
             }
         })();
-    }, [attributionId, location.search]);
+    }, [attributionId, location.pathname, location.search, refreshSubscriptionDetails, usePublicApiTeamsService]);
 
-    useEffect(() => {
-        setPendingStripeSubscription(undefined);
-        if (!attributionId) {
-            return;
-        }
-        try {
-            const pendingStripeSubscription = window.localStorage.getItem(localStorageKey);
-            if (!pendingStripeSubscription) {
-                return;
-            }
-            const pending = JSON.parse(pendingStripeSubscription);
-            setPendingStripeSubscription(pending);
-        } catch (error) {
-            console.warn("Could not load pending Stripe subscription", attributionId, error);
-        }
-    }, [attributionId, localStorageKey]);
-
-    useEffect(() => {
-        if (!pendingStripeSubscription || !attributionId) {
-            return;
-        }
-        if (!!stripeSubscriptionId) {
-            // The upgrade was successful!
-            window.localStorage.removeItem(localStorageKey);
-            clearTimeout(pollStripeSubscriptionTimeout!);
-            setPendingStripeSubscription(undefined);
-            return;
-        }
-        if (pendingStripeSubscription.pendingSince + 1000 * 60 * 5 < Date.now()) {
-            // Pending Stripe subscription expires after 5 minutes
-            window.localStorage.removeItem(localStorageKey);
-            clearTimeout(pollStripeSubscriptionTimeout!);
-            setPendingStripeSubscription(undefined);
-            return;
-        }
-        if (!pollStripeSubscriptionTimeout) {
-            // Refresh Stripe subscription in 5 seconds in order to poll for upgrade confirmation
-            const timeout = setTimeout(async () => {
-                await refreshSubscriptionDetails(attributionId);
-                setPollStripeSubscriptionTimeout(undefined);
-            }, 5000);
-            setPollStripeSubscriptionTimeout(timeout);
-        }
-    }, [
-        pendingStripeSubscription,
-        pollStripeSubscriptionTimeout,
-        stripeSubscriptionId,
-        attributionId,
-        localStorageKey,
-    ]);
-
-    useEffect(() => {
-        if (!attributionId) {
-            return;
-        }
-        getGitpodService().server.getUsageBalance(attributionId).then(setCurrentUsage);
-    }, [attributionId, billingCycleFrom]);
-
-    const showSpinner = !attributionId || isLoadingStripeSubscription || !!pendingStripeSubscription;
+    const showSpinner = !attributionId || isLoadingStripeSubscription || !!getPendingStripeSubscription(attributionId);
```

Comment:
It is unfortunately normal that it takes some time. I'll investigate if there is unnecessary waiting happening. But for me the spinning ends after a few seconds (without a reload). If this is all spent on the interaction with Stripe and we cannot improve it, we might consider making the 'spinning' more exciting.

Reply:
Found the likely issue: When the `subscriptionid` was successfully received it was only stored in localStorage but not in a react state, so there was no trigger for a re-rendering. I looked at the entire logic once more and figured you were right in that we can get rid of the localStorage state, because the state is created after the redirect. Much simpler now :orange_heart: